components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/GeneListController.java

```diff
@@ -183,4 +189,78 @@ public void writeJSON(Patient patient, JSONObject json, Collection<String> selec
             }
         }
     }
+
+    @Override
+    public PatientData<Map<String, String>> readJSON(JSONObject json)
+    {
+        Object testGeneJson = json.get(this.getJsonPropertyName());
+        if (testGeneJson == null) {
+            return null;
+        }
+
+        try {
+            JSONArray genesJson = json.getJSONArray(this.getJsonPropertyName());
+            List<Map<String, String>> allGenes = new LinkedList<Map<String, String>>();
+            for (Object geneJsonUncast : genesJson) {
+                JSONObject geneJson = JSONObject.fromObject(geneJsonUncast);
+                Map<String, String> singleGene = new LinkedHashMap<String, String>();
+                for (String property : getProperties()) {
+                    // todo. it makes no sense to have no gene name, but to have comments, which can occur
+                    if (geneJson.has(property)) {
+                        String field = geneJson.getString(property);
+                        if (field != null) {
+                            singleGene.put(property, field);
+                        }
+                    }
+                }
+                allGenes.add(singleGene);
+            }
+
+            if (allGenes.isEmpty()) {
+                return null;
+            } else {
+                return new IndexedPatientData<Map<String, String>>(getName(), allGenes);
+            }
+        } catch (Exception e) {
+            this.logger.error("Could not load genes from JSON", e.getMessage());
+        }
+        return null;
+    }
+
+    @Override
+    public void save(Patient patient)
+    {
+        try {
+            XWikiDocument doc = (XWikiDocument) this.documentAccessBridge.getDocument(patient.getDocument());
+            if (doc == null) {
+                throw new NullPointerException(ERROR_MESSAGE_NO_PATIENT_CLASS);
+            }
+
+            // todo. does not check for existing genes
+            PatientData<Map<String, String>> genes = patient.getData(getName());
+            if (!genes.isIndexed()) {
+                return;
+            }
+
+            XWikiContext context = (XWikiContext) this.execution.getContext().getProperty("xwikicontext");
+            Iterator<Map<String, String>> iterator = genes.iterator();
+            while (iterator.hasNext()) {
+                Map<String, String> gene = iterator.next();
```

Comment:
As our previous struggles with items in a list show, maybe we should add a try-catch within the while(), so that we catch and omit any individual mis-formatted/broken-in-any-way genes individually, without failing the entire gene import? Genes will be a complicated entity, so we are somewhat likely to fail import of amn individual gene even if all other genes are ok?